core/controller/src/main/scala/whisk/core/controller/Actions.scala

```diff
@@ -781,6 +796,318 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
         }
     }
 
+    /** Begin Sequence invocation/activation code */
+
+    /**
+     * Executes a sequence by invoking in a blocking fashion each of its components.
+     *
+     * @param user the user invoking the action
+     * @param action the sequence action to be invoked
+     * @param env the merged parameters from the package/reference if any TODO: check ignoring these params
+     * @param payload the dynamic arguments for the activation
+     * @param blocking true iff this is a blocking invoke
+     * @param topmost true iff this is the topmost sequence invoked directly through the api (not indirectly through a sequence)
+     * @param components the actions in the sequence
+     * @param cause the id of the activation that caused this sequence (defined only for inner sequences and None for topmost sequences)
+     * @param atomicActionsCount the dynamic atomic action count observed so far since the start of invocation of the topmost sequence(0 if topmost)
+     * @param transid a transaction id for logging
+     * @return a future of type (ActivationId, Some(WhiskActivation), atomicActionsCount) if blocking; else (ActivationId, None, 0)
+     */
+    private def invokeSequence(
+            user: Identity,
+            action: WhiskAction,
+            //env: Option[Parameters],
+            payload: Option[JsObject],
+            blocking: Boolean,
+            topmost: Boolean,
+            components: Vector[FullyQualifiedEntityName],
+            cause: Option[ActivationId],
+            atomicActionsCount: Int)(implicit transid: TransactionId): Future[(ActivationId, Option[WhiskActivation], Int)] = {
+        // create new activation id that corresponds to the sequence
+        val seqActivationId = activationIdFactory.make()
+        info(this, s"Invoking sequence $action topmost $topmost activationid '$seqActivationId'")
+        val start = Instant.now(Clock.systemUTC())
+        var seqActivation: Option[WhiskActivation] = None
```

Comment:
this `var` is potentially dangerous as it's racy wrt the future below that assigns it.